app/flatpak-builtins-permission-set.c

```diff
@@ -0,0 +1,247 @@
+ *
+ * Authors:
+ *       Matthias Clasen <mclasen@redhat.com>
+ */
+
+#include "config.h"
+
+#include <locale.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+#include <errno.h>
+
+#include <glib/gi18n.h>
+
+#include "libglnx/libglnx.h"
+#include "flatpak-permission-dbus-generated.h"
+
+#include "flatpak-builtins.h"
+#include "flatpak-table-printer.h"
+#include "flatpak-utils-private.h"
+#include "flatpak-run-private.h"
+
+static GOptionEntry options[] = {
+  { NULL }
+};
+
+static char **
+get_ids_for_table (XdpDbusPermissionStore *store,
+                   const char             *table)
+{
+  char **ids = NULL;
+
+  xdp_dbus_permission_store_call_list_sync (store, table, &ids, NULL, NULL);
+
+  return ids;
+}
+
+static const char *device_perms[] = { "yes", "no", "ask", NULL };
+static const char *document_perms[] = { "read", "write", "delete", "grant-permissions", NULL };
+static const char *notification_perms[] = { "yes", "no", NULL };
+
+static const char **
+get_permission_values_for_table (const char *table)
+{
+  if (strcmp (table, "devices") == 0)
+    return device_perms;
+  else if (strcmp (table, "documents") == 0)
+    return document_perms;
+  else if (strcmp (table, "notifications") == 0)
+    return notification_perms;
+
+  return NULL;
+}
+
+static gboolean
+set_permissions (XdpDbusPermissionStore  *store,
+                 const char              *table,
+                 const char              *id,
+                 const char              *app_id,
+                 const char             **permissions,
+                 GError                 **error)
+{
+  GVariant *perms = NULL;
+  GVariant *data = NULL;
+  GVariantBuilder builder;
+  int i;
+  GVariant *new_perms = g_variant_new_strv (permissions, -1);
+
+  if (!xdp_dbus_permission_store_call_lookup_sync (store, table, id,
+                                                   &perms, &data,
+                                                   NULL, error))
+    return FALSE;
+
+  g_variant_builder_init (&builder, G_VARIANT_TYPE ("a{sas}"));
+  for (i = 0; perms && i < g_variant_n_children (perms); i++)
+    {
+      const char *key;
+      GVariant *value = NULL;
+
+      g_variant_get_child (perms, i, "{&s@as}", &key, &value);
+      if (strcmp (key, app_id) == 0)
+        {
+          if (permissions[0] != NULL)
+            g_variant_builder_add (&builder, "{s@as}", key, new_perms);
+          new_perms = NULL;
+        }
+      else
+        g_variant_builder_add (&builder, "{s@as}", key, value);
+    }
+
+  if (new_perms != NULL)
+    {
+      if (permissions[0] != NULL)
+        g_variant_builder_add (&builder, "{s@as}", app_id, new_perms);
+    }
+
+  if (!xdp_dbus_permission_store_call_set_sync (store, table, TRUE, id,
+                                                g_variant_builder_end (&builder),
+                                                data ? data : g_variant_new_byte (0),
+                                                NULL, error))
```

Comment:
This should add the permissions for the app keeping the previous ones, so I'm not sure why you read the previous values out above? Also, seems like you can use the SetPermissions call which is easier if your just doing one app.

Reply:
true. I'll redo this